src/java.base/share/classes/jdk/internal/foreign/abi/aarch64/CallArranger.java

```diff
@@ -183,8 +188,19 @@ public StorageCalculator(boolean forArguments) {
 
         VMStorage stackAlloc(long size, long alignment) {
             assert forArguments : "no stack returns";
-            alignment = Math.max(alignment, STACK_SLOT_SIZE);
-            stackOffset = Utils.alignUp(stackOffset, alignment);
+            // Implementation limit: each arg must take up at least an 8 byte stack slot (on the Java side)
+            // There is currently no way to address stack offsets that are not multiples of 8 bytes
+            // The VM can only address multiple-of-4-bytes offsets, which is also not good enough for some ABIs
+            // see JDK-8283462 and related issues
+            long stackSlotAlignment = Math.max(alignment, STACK_SLOT_SIZE);
```

Comment:
Reading this fix makes me wonder a bit. In reality size and alignment are determined by the layout. So a user could, in principle, say "JAVA_INT.withBitAlignment(64)" and this check would be satisfied. I guess what I'm trying to get at is that the set of allowed size/alignment combos is set by the platform ABI. For instance, SysV specifies how to pass a 4-byte long, 4-byte aligned, little-endian `int` - but does not specify what happens if e.g. the alignment constraints of the value are different from that described in the ABI document. Now, in the case of SysV, the only case where alignment matters is for struct classification (structs that have unaligned fields are always passed in memory). But in the case of the MacOS ABI, altering the layout alignment of one or more value can result in subtle differences in how the function is invoked - and if the alignment constraint specified by the user does not agree with that of the ABI, while the linker will do _something_, it will almost certainly lead to an error.

So, I believe that, in addition to this, for all the ABIs that have special alignment-dependent treatment, we should apply some checks, to make sure that the layouts in the function descriptor conforms to the alignments specified in the ABI. What do you think?

Reply:
Yes that makes sense, good catch!

We need to do the same on x64 platforms, since we use the same logic to align stack values. So, for instance, someone could have a hyper-aligned `long` aligned to 128 bits, which would mean it is placed at the wrong point on the stack.

Reply:
Looking more at this. I think we already solve this by manually recomputing the alignment in `SharedUtils.alignment` called from the other overload of this method.

Reply:
I think `SharedUtils::alignment` overrides whatever alignment constraint is specified in the layout. This function has existed for a very long time (probably from before layouts even supported alignment constraints). Some pieces of this code seems to be influenced by how SysV handles alignment constraints - especially the array alignment implementation.
IMHO, an ABI should check that the incoming layouts have the correct alignment constraints, rather than drop the constraints, and then recomputing them from scratch.

Reply:
Yes, I agree. I think in the past that would have meant that users would always have to adjust the alignment of JAVA_* layouts before being able to link with them, but not that those layouts have the right alignment constraints we can more explicitly reject layouts that have the wrong alignment during linking.

Reply:
I've filed a followup for this: https://bugs.openjdk.java.net/browse/JDK-8284300